### ISSUE TITLE
Add streaming to Remote Read endpoint

### ISSUE THREAD

**Issue by gouthamve** [prometheus/prometheus#4517]
Remote read is quite inefficient right now and could be made much better through streaming. This has been discussed in the dev-summit: https://docs.google.com/document/d/1-C5PycocOZEVIPrmM1hn8fBelShqtqiAmFptoG4yK70/edit#heading=h.5d2mc4i0bnnp

**Comment by Bplotka** [comment]
I can take it if no one started working on this.

Design doc: https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit?usp=sharing

Dev-mailing topic: https://groups.google.com/forum/#!topic/prometheus-developers/79b6j_6F_Vc

**Comment by harshal-shah** [comment]
According to https://github.com/prometheus/prometheus/pull/4655 it seems this has been resolved in prometheus 2.5.0

**Comment by rshivane** [comment]
I would like to add that, as a vendor implementing the server-side of remote read API, the lack of streaming adds similar memory pressure on the marshaling side.

@bwplotka I notice that your already have a PR for streaming support in #4591. Whenever this is approved for merge, we would be happy to implement it on our storage and validate.

**Comment by bwplotka** [comment]
Yes. This is working code for streaming with raw samples, but we have some internal discussion how it should be done. I closed it to reduce confusion for now: https://github.com/prometheus/prometheus/pull/4591

TL;DR: It would be nice to have remote read stream with Prometheus chunk encoded samples (instead of row samples), to avoid unnecessary decoding & encoding. Does it 
The plan is to:
1. Change the doc to propose this idea (right now it proposes something in between)
2. Setup a call to gather feedback on this idea.

Did not have much time to jump into it yet, but I am happy to hear that you are ok to help with this!

**Comment by rshivane** [comment]
>  It would be nice to have remote read stream with Prometheus chunk encoded samples (instead of row samples), to avoid unnecessary decoding & encoding.

Is that essentially ["Alternative 1" in the design doc](https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit#heading=h.12fpmrn0ayj) ?

**Comment by bwplotka** [comment]
Exactly, but need to reorder ;p

**Comment by bwplotka** [comment]
Just FYI, since we got concensus on this, expect more finalized design doc (same link: https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit#) and initial benchmarks soon. I scheduled some time for this work in June.

**Comment by bwplotka** [comment]
Back to business!

Protobuf changes: https://github.com/prometheus/prometheus/pull/5701
Delimited streaming: https://github.com/prometheus/prometheus/pull/5702
All, including basic server side implementation (draft): https://github.com/prometheus/prometheus/pull/5703
Caller usage example (Thanos): https://github.com/improbable-eng/thanos/pull/1268 

Living document: https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit#

**Comment by bwplotka** [comment]
First MVP should be safe to use. I build 2 docker images if someone wants to try:
* `bplotka/prometheus:v2.11.0-rc.0-rr-streaming` with remote read streaming server support (https://github.com/prometheus/prometheus/pull/5703)
* `improbable/thanos:v0.5.0-rr-streamed2` with thanos sidecar remote read streaming support.

No config changes, just move docker images to that Prometheus and Thanos sidecar tags and *enjoy*!

2 big requests against Prometheus + sidecar in upgraded version look like this memory and CPU wise:
![image](https://user-images.githubusercontent.com/6950331/60721596-62ff8f80-9f26-11e9-808d-cc388b6666f6.png)

Versus old one with non-streaming sampled read (3GB of mem!):
![image](https://user-images.githubusercontent.com/6950331/60721716-b7a30a80-9f26-11e9-8822-3c18a7e764b0.png)

**Comment by bwplotka** [comment]
Remote read now supports streaming.

**Comment by bwplotka** [comment]
Follow up tasks:
* Docs: https://github.com/prometheus/prometheus/issues/5925
* Chunk iterator instead of reencoding:  https://github.com/prometheus/prometheus/pull/5882
* https://github.com/prometheus/prometheus/issues/5926
